src/rgw/rgw_dedup_cluster.cc

```diff
@@ -0,0 +1,1068 @@
+		  << num_work_shards << "\n" << std::endl;
+      }
+
+      for (unsigned shard = 0; shard < num_work_shards; shard++) {
+	if (bl_arr[shard].length() == 0) {
+	  display_progress(sp_arr[shard].progress_a, sp_arr[shard].progress_b, shard);
+	  continue;
+	}
+	completed_work_shards_count++;
+	worker_stats_t stats;
+	try {
+	  auto p = bl_arr[shard].cbegin();
+	  decode(stats, p);
+	  wrk_stats_sum += stats;
+	}catch (const buffer::error&) {
+	  std::cerr << __func__ << "::(2)failed worker_stats_t decode #" << shard << std::endl;
+	  continue;
+	}
+	collect_single_shard_stats(dpp, owner_map, sp_arr, shard, &show_time, "WORKER");
+      }
+      std::cout << "Aggreagted work-shard stats counters:\n" << wrk_stats_sum << std::endl;
+      md5_start_time = show_time_func(epoch.time, show_time, owner_map);
+    }
+
+    if (completed_work_shards_count == num_work_shards) {
+      std::map<std::string, member_time_t> owner_map;
+      bool show_time = true;
+      md5_stats_t md5_stats_sum;
+      bufferlist bl_arr[num_md5_shards];
+      shard_progress_t sp_arr[num_md5_shards];
+      int cnt = collect_shard_stats(ioctx, dpp, epoch.time, num_md5_shards,
+				    MD5_SHARD_PREFIX, bl_arr, sp_arr);
+      if (cnt != num_md5_shards) {
+	std::cerr << ">>>Partial MD5_SHARD stats recived " << cnt << " / "
+		  << num_md5_shards << "\n" << std::endl;
+      }
+
+      for (unsigned shard = 0; shard < num_md5_shards; shard++) {
+	if (bl_arr[shard].length() == 0) {
+	  display_progress(sp_arr[shard].progress_a, sp_arr[shard].progress_b, shard);
+	  continue;
+	}
+	completed_md5_shards_count++;
+	md5_stats_t stats;
+	try {
+	  auto p = bl_arr[shard].cbegin();
+	  decode(stats, p);
+	  md5_stats_sum += stats;
+	}catch (const buffer::error&) {
+	  std::cerr << __func__ << "::failed md5_stats_t decode #" << shard << std::endl;
+	  continue;
+	}
+	collect_single_shard_stats(dpp, owner_map, sp_arr, shard, &show_time, "MD5");
+      }
+      std::cout << "Aggreagted md5-shard stats counters:\n" << md5_stats_sum << std::endl;
+      show_dedup_ratio(wrk_stats_sum, md5_stats_sum);
+      show_time_func(md5_start_time, show_time, owner_map);
+    }
+
+    if (completed_md5_shards_count == num_md5_shards) {
+      std::cout << "DEDUP WORK WAS COMPLETED" << std::endl;
+    }
+    return 0;
+  }
+
+  //---------------------------------------------------------------------------
+  // command-line called from radosgw-admin.cc
+  int cluster::dedup_control(rgw::sal::RadosStore *store,
+			     const DoutPrefixProvider *dpp,
+			     int urgent_msg)
+  {
+    ldpp_dout(dpp, 20) << __func__ << "::dedup_control req = "
+		       << get_urgent_msg_names(urgent_msg) << dendl;
+    if (urgent_msg != URGENT_MSG_RESUME  &&
+	urgent_msg != URGENT_MSG_PASUE   &&
+	urgent_msg != URGENT_MSG_RESTART &&
+	urgent_msg != URGENT_MSG_ABORT) {
+      ldpp_dout(dpp, 1) << __func__ << "::illegal urgent_msg="<< urgent_msg << dendl;
+      return -1;
```

Comment:
change to `-EINVAL`.
also would probably be better to use an enum here

Reply:
done